### PR TITLE
Core: Add property to prevent loss of view representation when replacing a view

### PR DESCRIPTION
**Pull request by nastra**
The general idea behind this approach is to prevent a `CREATE OR REPLACE VIEW` stmt in Trino/Spark from accidentally dropping a view that was e.g. created in Spark/Trino. 

In Spark, the strictness flag could be disabled via `ALTER VIEW ... SET TBLPROPERTIES ('replace.strictness.enabled' = 'false')` or during a replace via `CREATE OR REPLACE VIEW ... TBLPROPERTIES ('replace.strictness.enabled' = 'false')`.

The PR still needs more tests, which will be added once we decide on the direction going forward

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/rest/RESTViewOperations.java [hunk 1]
```diff
@@ -59,6 +60,8 @@ public void commit(ViewMetadata base, ViewMetadata metadata) {
     // this is only used for replacing view metadata
     Preconditions.checkState(base != null, "Invalid base metadata: null");
 
+    ViewUtil.verifyReplaceStrictness(base, metadata);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTViewOperations.java:63]: Could this check be done in `ViewMetadata.Builder` instead? That seems like a safer place because it is used by all view catalog implementations.

File: core/src/main/java/org/apache/iceberg/view/ViewProperties.java [hunk 2]
```diff
@@ -26,6 +26,8 @@ public class ViewProperties {
   public static final String METADATA_COMPRESSION = "write.metadata.compression-codec";
   public static final String METADATA_COMPRESSION_DEFAULT = "gzip";
   public static final String COMMENT = "comment";
+  public static final String REPLACE_STRICTNESS_ENABLED = "replace.strictness.enabled";
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/view/ViewProperties.java:29]: I think this is over-broad. There could be many forms of "strict" replacement. For example, it is more strict to ensure that the column names and positions don't change in a replace operation, but that's not what we're addressing here.

How about something like `replace.drop-dialect.allowed=false`? That would prevent the caller from dropping a view dialect.

File: core/src/main/java/org/apache/iceberg/rest/RESTViewOperations.java (hunk 1 above)
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/rest/RESTViewOperations.java:63]: What's the rationale for having this in the `ViewUtil` and in a public API? I think it makes more sense to be `protected` within `ViewOperations` so that different implementations can just call out to it during the commit. I don't think it needs to be exposed more than that at least at this point in time.

**Review by amogh-jahagirdar — COMMENTED** [review]
Thanks, I think this makes sense just want to make sure we think through alternatives as well.

The main alternative I can think of is that the view implementation simply takes a stance on behavior instead of a property which is essentially what we had before (we'd allow a representation do be dropped during a replace). Of course users may not want this as a replace from one system (which wasn't coordinated with the other) would lead to failed workloads on the other system. In case they do want to coordinate this, and perform the replace from the other system, they can update the property and perform the replace. The opposite stance of always failing doesn't make much sense since it removes the utility of being able to access Iceberg views from different engines.

So I do think this approach makes sense. The only other comment I have is the same that @rdblue pointed out which is what's being done here is a more granular strictness check on the dialects. I'd propose a property called `replace.dialect-loss.allowed`

File: core/src/main/java/org/apache/iceberg/view/ViewProperties.java (hunk 2 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/view/ViewProperties.java]: I agree that the inital naming was quite broad. I've updated this to `replace.drop-dialect.allowed` with `false` being the default

File: core/src/main/java/org/apache/iceberg/rest/RESTViewOperations.java (hunk 1 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/RESTViewOperations.java]: see https://github.com/apache/iceberg/pull/9620#discussion_r1478246464

File: core/src/main/java/org/apache/iceberg/view/ViewMetadata.java [hunk 3]
```diff
@@ -510,5 +521,29 @@ static List<ViewHistoryEntry> updateHistory(List<ViewHistoryEntry> history, Set<
...
       return changes.stream().filter(updateClass::isInstance).map(updateClass::cast);
     }
+
+    private void checkIfDialectIsDropped(ViewVersion previous, ViewVersion current) {
+      Set<String> baseDialects = sqlDialectsFor(previous);
+      Set<String> updatedDialects = sqlDialectsFor(current);
+
+      Preconditions.checkState(
+          updatedDialects.containsAll(baseDialects),
+          "Cannot replace view due to loss of view representations (%s=false):\nPrevious representations: %s\nNew representations: %s",
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:531]: "representations" should be "dialects" here.

File: core/src/main/java/org/apache/iceberg/view/ViewMetadata.java [hunk 4]
```diff
@@ -450,6 +452,15 @@ public ViewMetadata build() {
           ViewProperties.VERSION_HISTORY_SIZE,
           historySize);
 
+      if (!PropertyUtil.propertyAsBoolean(
+          properties,
+          ViewProperties.REPLACE_DROP_DIALECT_ALLOWED,
+          ViewProperties.REPLACE_DROP_DIALECT_ALLOWED_DEFAULT)) {
+        for (int i = Math.max(1, startingViewVersionsSize); i < versions.size(); i++) {
+          checkIfDialectIsDropped(versions.get(i - 1), versions.get(i));
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:460]: This check is overbroad. This check should not cause a view to be completely broken just because a different client put the view into a bad state. Instead, this check should only run to validate the changes made by the caller.

This also makes a bad assumption about view versions. Consecutive view version numbers have no meaning. The current view version may not be the highest number, since the current version ID is tracked separately.

This PR needs to validate that when the current view version changes, there are no dropped dialects.

Comment (nastra) [reply core/src/main/java/org/apache/iceberg/view/ViewMetadata.java]: indeed this was a bad assumption on my end. We need to check against the view versions that have been added by the client, which I've changed now and also added a test so that things fail correctly when the view version id is being set by the client

File: core/src/main/java/org/apache/iceberg/view/ViewMetadata.java [hunk 5]
```diff
@@ -510,5 +529,29 @@ static List<ViewHistoryEntry> updateHistory(List<ViewHistoryEntry> history, Set<
...
+          baseDialects,
+          updatedDialects);
+    }
+
+    private Set<String> sqlDialectsFor(ViewVersion viewVersion) {
+      Set<String> dialects = Sets.newHashSet();
+      for (ViewRepresentation repr : viewVersion.representations()) {
+        if (repr instanceof SQLViewRepresentation) {
+          SQLViewRepresentation sql = (SQLViewRepresentation) repr;
+          dialects.add(sql.dialect());
```
[the remark this answers is not in the document]
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:560]: When detecting whether we have multiple dialects in https://github.com/apache/iceberg/blob/d2be90d9cf087b7f644350dd251071e15ff71a11/core/src/main/java/org/apache/iceberg/view/ViewMetadata.java#L295 we use the dialect as-is, hence I'm doing the same here.
I think we should be ok not doing a `toLowerCase()` in both of these cases as typically an engine would add & read a dialect using the same casing

File: core/src/test/java/org/apache/iceberg/view/TestViewMetadata.java [hunk 6]
```diff
@@ -334,6 +334,9 @@ public void viewMetadataAndMetadataChanges() {
             .addVersion(viewVersionOne)
             .addVersion(viewVersionTwo)
             .addVersion(viewVersionThree)
+            .setCurrentVersionId(1)
+            .setCurrentVersionId(2)
+            // this will only keep the last set version id and ignore the previously set ones
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/view/TestViewMetadata.java:339]: this test already verifies that there's only a single `SetCurrentViewVersion` in `changes`

File: core/src/main/java/org/apache/iceberg/view/ViewMetadata.java [hunk 7]
```diff
@@ -479,6 +491,23 @@ public ViewMetadata build() {
           metadataLocation);
     }
 
+    private void removeIntermediateViewVersionIdChanges() {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:494]: It looks like I was wrong about the way `TableMetadata` works. I don't think it is rewriting the set of changes in the build method after all. I think we should do that as well to keep it simple and avoid a few problems with doing it this way (`removeIf` will not remove duplicates if they set the same version as current).

File: core/src/main/java/org/apache/iceberg/view/ViewMetadata.java [hunk 8]
```diff
@@ -479,6 +491,23 @@ public ViewMetadata build() {
...
+        // this can't compare SetCurrentViewVersion.version() with currentVersionId as
+        // SetCurrentViewVersion.version() can be -1 (to indicate that the versionId can be
+        // re-assigned on the server)
+        changes.removeIf(
+            c ->
+                (c instanceof MetadataUpdate.SetCurrentViewVersion)
+                    && !c.equals(lastSetViewVersionId));
+      }
+    }
+
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:510]:
```suggestion
```

File: core/src/main/java/org/apache/iceberg/view/ViewMetadata.java [hunk 9]
```diff
@@ -438,6 +440,16 @@ public ViewMetadata build() {
           metadataLocation == null || changes.isEmpty(),
           "Cannot create view metadata with a metadata location and changes");
 
+      removeIntermediateViewVersionIdChanges();
+
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:444]:
```suggestion
```

File: core/src/main/java/org/apache/iceberg/view/ViewMetadata.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:560]: I agree that dialect should not be case sensitive. We should fix it in both places.

File: core/src/test/java/org/apache/iceberg/view/TestViewMetadata.java (hunk 6 above)
Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/view/TestViewMetadata.java:339]: I don't think we need this if we allow duplicates.

File: core/src/test/java/org/apache/iceberg/view/TestViewMetadata.java [hunk 10]
```diff
@@ -799,4 +802,299 @@ public void lastAddedSchemaFailure() {
...
+  public void droppingDialectDoesNotFailWhenAllowed() {
+    Schema schema = new Schema(Types.NestedField.required(1, "x", Types.LongType.get()));
+    ViewRepresentation spark =
+        ImmutableSQLViewRepresentation.builder().dialect("spark").sql("select * from tbl").build();
+    ViewRepresentation trino =
+        ImmutableSQLViewRepresentation.builder().dialect("trino").sql("select * from tbl").build();
+
+    ViewMetadata metadata =
+        ViewMetadata.builder()
+            .setLocation("custom-location")
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/view/TestViewMetadata.java:859]: Why is this needed?

**Review by rdblue — APPROVED** [review]
Looks good to me. I think we should update this slightly to remove the part that drops intermediate `SetCurrentViewVersion` changes, since we don't do it for tables. (And would have to fix the implementation.)

File: core/src/test/java/org/apache/iceberg/view/TestViewMetadata.java (hunk 10 above)
Comment (nastra) [reply core/src/test/java/org/apache/iceberg/view/TestViewMetadata.java:859]: `location` is a required field, thus it has to be set

File: core/src/main/java/org/apache/iceberg/view/ViewMetadata.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/view/ViewMetadata.java]: I've updated this in both places and updated the test to check for this

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Merged! Thanks for getting this done for the 1.5 release, @nastra!
